pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java

```diff
@@ -0,0 +1,152 @@
+        canalSourceConfig = CanalSourceConfig.load(config);
+        if (canalSourceConfig.getCluster()) {
+            connector = CanalConnectors.newClusterConnector(canalSourceConfig.getZkServers(),
+                    canalSourceConfig.getDestination(), canalSourceConfig.getUsername(), canalSourceConfig.getPassword());
+        } else {
+            connector = CanalConnectors.newSingleConnector(
+                    new InetSocketAddress(canalSourceConfig.getSingleHostname(), canalSourceConfig.getSinglePort()),
+                    canalSourceConfig.getDestination(), canalSourceConfig.getUsername(), canalSourceConfig.getPassword());
+        }
+        log.info("start canal connect");
+        this.start();
+
+    }
+
+    protected void start() {
+        Assert.notNull(connector, "connector is null");
+        thread = new Thread(new Runnable() {
+
+            @Override
+            public void run() {
+                process();
+            }
+        });
+
+        thread.setUncaughtExceptionHandler(handler);
+        running = true;
+        thread.start();
+    }
+
+    @Override
+    public void close() throws InterruptedException {
+        log.info("close canal source");
+        if (!running) {
+            return;
+        }
+        running = false;
+        if (thread != null) {
+            thread.interrupt();
+            thread.join();
+        }
+        if (connector != null) {
+            connector.disconnect();
+        }
+
+        MDC.remove("destination");
+    }
+
+    protected void process() {
+        while (running) {
+            try {
+                MDC.put("destination", canalSourceConfig.getDestination());
+                connector.connect();
+                log.info("start canal process");
+                connector.subscribe();
+                while (running) {
+                    Message message = connector.getWithoutAck(canalSourceConfig.getBatchSize());
+                    // delete the setRaw in new version of canal-client
+                    message.setRaw(false);
+                    List<FlatMessage> flatMessages = FlatMessage.messageConverter(message);
+                    long batchId = message.getId();
+                    int size = message.getEntries().size();
+                    if (batchId == -1 || size == 0) {
+                        try {
+                            Thread.sleep(1000);
+                        } catch (InterruptedException e) {
+                        }
+                    } else {
+                        if (flatMessages != null) {
+                            for (FlatMessage flatMessage : flatMessages) {
+                                String m = JSON.toJSONString(flatMessage, SerializerFeature.WriteMapNullValue);
+                                consume(new CanalRecord(m.getBytes(), batchId));
+                            }
+                        }
+                    }
+
+                    connector.ack(batchId);
```

Comment:
if you are acking here, it only supports at-most-once semantic. you might consider implementing ack() in CanalRecord